### DISCUSSION TITLE
Why does this CH340G adapter work so well, with such a crappy design?

### DISCUSSION THREAD

**Discussion by hmeijdam** [SpenceKonde/megaTinyCore#1189] [Q&A]
I am using this CH340G adapter for all my Serial UPDI programming for a few years now. It works just great on all speeds including TURBO speed.

But Why?

It violates all design recommendations like:
"Use a series resistor somewhere between the CH340G and the UPDI pin". Not this one.
"Never use any LED connected to TX or RX". This one has two LEDs, and they even skimped on one series resistor.
"Use a schottky diode between RX and TX". I seemed to have plucked a signal diode from the parts bin with a 0.7V Forward Voltage.

Here is the schematic of the connection between the CH340G and the Attiny.
![image](https://github.com/user-attachments/assets/c23788c3-7210-4a87-8416-64a874c5b703)

I bought them 5pcs for 2 Euro's a few years back, and expected them to miserably fail, but they became my go-to programmer in the end.

**Comment by pfoun** [comment]
I don't know the answer, but here is my design with a CH340N, and it works very well.  More reliable than the ones I had made with the CP2102.  I'm suprised you did not suggest a CH340 in your recent post about UPDI with the CP2102.  
![20241230_101500](https://github.com/user-attachments/assets/0392b1a5-6603-4c2d-94ca-58d348c7151f)

**Comment by SpenceKonde** [comment]
That the design works is a miracle, frankly, they do so many stupid things (that power voltage switch they use switches it directly... It's rated for tens of milliamps. People routinely expect to pull far more than that.)

I never even tried using them with the idiodically connected LED enabled, Really, the RX LED should prevent UPDI from working. You using it at 3V3?. At that point, with RX tied to ground, the voltage across the 1k resistor might be less than 1 volt due to the forward drop across the LED, hence less than 1 ma of current, and to the extent that the voltage on Rx was above ground, would lower that further, such that the LED's influence is negligible as the voltage on RX approached levels that would no longer register as LOW.
